File: Patterns/Command Pattern/Cpp/Command/GarageDoor/GarageDoorOpenCommand.cpp
#include "../ICommand.cpp"
#include "./GarageDoor.cpp"

class GarageDoorOpenCommand : public ICommand {
private:
    GarageDoor* garageDoor;
    int preview;

public:
    GarageDoorOpenCommand(GarageDoor* garageDoor) : garageDoor{garageDoor}, preview{garageDoor->OFF} {}

    void execute() override {
        preview = garageDoor->getState();
        garageDoor->up();
        garageDoor->lightOn();
        garageDoor->stop();
    }

    void undo() override {
        if (preview == garageDoor->OFF) {
            garageDoor->down();
            garageDoor->lightOff();
            garageDoor->stop();
        }
    }

    string getCommandName() override {
        return "GarageDoorOpenCommand";
    }
};

File: Patterns/Command Pattern/CSharp/Command/GarageDoor/GarageDoorOpenCommand.cs
namespace Command.GarageDoor
{
    public class GarageDoorOpenCommand : ICommand
    {
        private GarageDoor garageDoor;
        private int preview;

        public GarageDoorOpenCommand(GarageDoor garageDoor)
        {
            preview = GarageDoor.OFF;
            this.garageDoor = garageDoor;
        }

        public void Execute()
        {
            preview = garageDoor.State;
            garageDoor.Down();
            garageDoor.LightOff();
            garageDoor.Stop();
        }

        public void Undo()
        {
            if (preview == GarageDoor.ON)
            {
                garageDoor.Up();
                garageDoor.LightOn();
                garageDoor.Stop();
            }
        }
    }
}